spec/grouping-nodes.md

```diff
@@ -0,0 +1,146 @@
+DataPlaneRole based grouping will not be automatically applied. So we need a
+different solution.
+
+Moreover the current restriction that a DataPlaneNode can only belong to a
+single DataPlaneRole is limiting. From nova perspective a set of compute nodes
+can be configured with CPU pinning while from neutron perspective a set of
+compute nodes can be configured with DPDK. If the two sets of nodes overlap but
+are not equal then there will be a need for additional roles to be defined for
+each combination of the configuration (pinning with DPDK, pinning without
+DPDK, sharing with DPDK, sharing without DPDK). This shows that the current
+restriction can lead to the explosion of the number of necessary groups to
+model the different combinations of service configs.
+
+## Proposed change
+
+Allow an EDPM node to belong to multiple groups by defining OpenStack service
+specific profiles. The DataPlaneRole could remain as is today but it would
+represent the grouping of the EDPM nodes only just from the generic
+hardware, host OS, infrastructure configuration perspective.
+
+Each service operator that has EDPM specific service CRD (e.g.
+NovaExternalCompute, OVNMetadataAgent) can define a Profile CRD (e.g.
+ComputeProfile, NetworkProfile) to describe the service specific configuration
+of a set of EDPM nodes. Then the human deployer can use annotations on the
+DataPlaneRole or DataPlaneNode to select a single Profile per service type.
+
+### Example
+
+ComputeProfile with shared CPUs
+```yaml
+apiVersion: nova.openstack.org/v1beta1
+kind: ComputeProfile
+metadata:
+    name: shared-cpus
+spec:
+    customComputeServiceConfig: |
+    [compute]
+    cpu_shared_set = 4-12,^8,15
+    [DEFAULT]
+    cpu_allocation_ratio = 4.0
+```
+
+ComputeProfile with pinned CPUs
+```yaml
+apiVersion: nova.openstack.org/v1beta1
+kind: ComputeProfile
+metadata:
+name: dedicated-cpus
+spec:
+    customComputeServiceConfig: |
+    [compute]
+    cpu_dedicated_set = 4-12,^8,15
+    [DEFAULT]
+    cpu_allocation_ratio = 1.0
+```
+
+OpenstackDataPlaneRole/edpm-compute
+```yaml
+apiVersion: dataplane.openstack.org/v1beta1
+kind: OpenStackDataPlaneRole
+metadata:
+    name: edpm-compute
+    annotations:
+        nova.openstack.org/cell: nova-cell1
+        nova.openstack.org/compute-profile: shared-cpus
+spec:
+...
+```
+This Role defines a set of EDPM compute nodes with shared CPUs.
+
+
+OpenstackDataPlaneNode/edpm-compute-0
+```yaml
+apiVersion: dataplane.openstack.org/v1beta1
+kind: OpenStackDataPlaneNode
+metadata:
+    name: edpm-compute-0
+spec:
+    role: edpm-compute
+...
+```
+This EDPM node is a compute that has shared CPUs as the Role selected the
+shared-cpu ComputeProfile
+
+OpenstackDataPlaneNode/edpm-compute-1
+```yaml
+apiVersion: dataplane.openstack.org/v1beta1
+kind: OpenStackDataPlaneNode
+metadata:
+    name: edpm-compute-1
+    annotations:
+        nova.openstack.org/compute-profile: pinned-cpus
+        neutron.openstack.org/network-profile: dpdk
```

Comment:
So to be clear here, you've defined this as having role `edpm-compute` but have overridden each thing defined by the role, presumably just because you have to have *some* role right? It might be a better example to only override one of these two, or add a third profile that you still inherit from the role to make it clear that you can compose them that way.

That said, I hope people don't have to do this too much and that it is reserved for special one-off special-purpose machines.

Reply:
I would not expect you would have to do this at the node level nessiarly but it would be good to have a multi-service example at the role level.

i.e. have OpenstackDataplaneRole edpm-nfv that uses the pinned-cpus nova profile and dpdk neutron profile.

that is how I would generally expect that to work.

with that said one use case for overriding at the node level is if the role had 
nova.openstack.org/compute-profile: <whatever>
telemetry.openstack.org/-profile: collector-rack1

but you wanted to use one of the nodes in that role to run extra services for that set of hosts

for example perhaps one of the nodes might also run collected/vector for monitoring aggreation

so that single node might have

nova.openstack.org/compute-profile: <whatever>
telemetry.openstack.org/-profile: collector-aggreator-rack1


that a bit contrived but the concept is simple all host in rack one would collect telemetry and forward it to the single node in rackone which is designated the aggrator that would then forward the telemetry for the entire rack to the central component.


unless you had something like that where all the node are basically the same expect one is slightly tweek i think it would make more sense to just have two dataplane Roles.

Reply:
I think your understanding is correct. The profile annotations in the node override the profile annotations inherited from the role.  Sean, you are probably right that it has a limited use case. Still because we have a Role - Node inheritance in the dataplane-operator we need to define what it means if the profile annotations are added to both the Node an the corresponding Role. So I choose the obvious meaning, the annotation on the more specific entity, the Node, overrides the annotation inherited from the more generic entity, the Role.

 I added the telemetry-profile to the example to show that we still inherit profiles from the Role that are note explicitly overridden on the Node.